### PR TITLE
feat: Allow creating multiple presigns in one TX

### PR DESCRIPTION
**Pull request by ItayLevyOfficial**
Until now, a user could only create one presign at a time. This PR introduce a `batch_size` parameter to the Move presign function, allowing users to create multiple presigns in the same TX.

Moreover, the presign output is now included in the batch completion event, so users don't need to fetch the presign object after catching the completion event.

### CODE REVIEW

File: crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move [hunk 1]
```diff
@@ -581,6 +635,7 @@ module pera_system::dwallet_2pc_mpc_ecdsa_k1 {
             });
             i = i + 1;
         };
+        vector::reverse(&mut message_approvals);
```
Comment (yaelAbergel) [inline crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move:638]: Why?

File: crates/pera-types/src/messages_dwallet_mpc.rs [hunk 2]
```diff
@@ -15,22 +15,30 @@ pub enum MPCRound {
     /// The first round of the Presign protocol.
     /// Contains the `ObjectId` of the dWallet object,
     /// and the DKG decentralized output.
-    PresignFirst(ObjectID, Vec<u8>),
+    PresignFirst(ObjectID, Vec<u8>, ObjectID),
     /// The second round of the Presign protocol.
     /// Contains the `ObjectId` of the dWallet object,
     /// and the Presign first round output.
-    PresignSecond(ObjectID, Vec<u8>),
+    PresignSecond(ObjectID, Vec<u8>, ObjectID),
```
Comment (yaelAbergel) [inline crates/pera-types/src/messages_dwallet_mpc.rs:22]: Update the documentation with the new parameter

File: sdk/typescript/src/dwallet-mpc/globals.ts [hunk 3]
```diff
@@ -139,3 +139,45 @@ export const getEventByTypeAndSessionId = async (
...
+	sessionID: string,
+	eventType: string,
+	isEventFn: (parsedJson: unknown) => parsedJson is TEvent,
+): Promise<TEvent> {
+	const startTime = Date.now();
+	let cursor = null;
+
+	while (Date.now() - startTime <= c.timeout) {
+		// Wait for a bit before polling again, objects might not be available immediately.
+		await new Promise((resolve) => setTimeout(resolve, 5_000));
```
Comment (yaelAbergel) [inline sdk/typescript/src/dwallet-mpc/globals.ts:154]: You can use the timeout from the config

File: crates/pera-core/src/dwallet_mpc/mpc_events.rs [hunk 4]
```diff
@@ -161,10 +173,8 @@ pub struct StartSignRoundEvent {
     pub(super) dkg_output: Vec<u8>,
     /// Hashed messages to Sign.
     pub(super) hashed_message: Vec<u8>,
-    /// Presign first round output, required for the MPC Sign session.
-    pub(super) presign_first_round_output: Vec<u8>,
-    /// Presign second round output, required for the MPC Sign session.
-    pub(super) presign_second_round_output: Vec<u8>,
+    /// The serialized presign object
```
Comment (yaelAbergel) [inline crates/pera-core/src/dwallet_mpc/mpc_events.rs:176]:
```suggestion
    /// The serialized final presign output, constructed from the outputs of both the first and second presign MPC rounds.
```

**Review by yaelAbergel — APPROVED** [review]
Just note my comments :)

File: crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move (hunk 1 above)
Comment (ItayLevyOfficial) [reply crates/pera-framework/packages/pera-system/sources/dwallet_2pc_mpc_ecdsa_k1.move:638]: Read the function code. At each iteration we pop the last message, so the message approvals are "reversed", until this line, in respect to the messages

File: sdk/typescript/src/dwallet-mpc/globals.ts (hunk 3 above)
Comment (ItayLevyOfficial) [reply sdk/typescript/src/dwallet-mpc/globals.ts:154]: No. we wait here five seconds before polling, not ten minutes (the timeout) before start polling.

File: Cargo.toml [hunk 5]
```diff
@@ -602,6 +601,7 @@ prometheus-closure-metric = { path = "crates/prometheus-closure-metric" }
 shared-crypto = { path = "crates/shared-crypto" }
 simulacrum = { path = "crates/simulacrum" }
 pera = { path = "crates/pera" }
+dwallet-mpc = { path = "crates/dwallet-mpc-centralized-party" }
```
Comment (zeevmoney) [inline Cargo.toml:604]: ???

Comment (ItayLevyOfficial) [reply Cargo.toml:604]: This dependency was not configured correctly as a member of this project, this line fixes this.

File: crates/pera-core/src/consensus_handler.rs [hunk 6]
```diff
@@ -438,24 +438,22 @@ impl<C: CheckpointServiceNotify + Send + Sync> ConsensusHandler<C> {
...
-                                        hashed_message,
-                                        output.clone(),
-                                    ) {
-                                        error!("Error storing message in batch: {:?}", err);
-                                        continue;
+                                    if let Err(err) = batches_manager
+                                        .store_verified_output(session_info.clone(), output.clone())
+                                    {
+                                        error!(
+                                            "Error storing verified output for session {:?}: {:?}",
```
Comment (zeevmoney) [inline crates/pera-core/src/consensus_handler.rs:452]:
```suggestion
                                            "error storing verified output for session {:?}: {:?}",
```
errors messages are always lower case

File: crates/pera-types/src/messages_dwallet_mpc.rs [hunk 7]
```diff
@@ -14,23 +14,31 @@ pub enum MPCRound {
...
     /// Contains the `ObjectId` of the dWallet object,
-    /// and the DKG decentralized output.
-    PresignFirst(ObjectID, Vec<u8>),
+    /// the DKG decentralized output, and the batch session ID.
+    PresignFirst(ObjectID, Vec<u8>, ObjectID),
     /// The second round of the Presign protocol.
     /// Contains the `ObjectId` of the dWallet object,
-    /// and the Presign first round output.
-    PresignSecond(ObjectID, Vec<u8>),
+    /// , the Presign first round output, and the batch session ID.
```
Comment (zeevmoney) [inline crates/pera-types/src/messages_dwallet_mpc.rs:21]:
```suggestion
    /// the Presign first round output, and the batch session ID.
```

File: sdk/typescript/test/e2e/dwallet-mpc.test.ts [hunk 8]
```diff
@@ -73,30 +73,49 @@ describe('Test dWallet MPC', () => {
...
+			.serialize([presignOutput1.presign, presignOutput2.presign])
+			.toBytes();
+		// let serializedPresignSessionIds = bcs
+		// 	.vector(bcs.string())
+		// 	.serialize([
+		// 		presignOutput1.first_round_session_id.slice(2),
+		// 		presignOutput2.first_round_session_id.slice(2),
+		// 	])
+		// 	.toBytes();
+		// TODO: undo this intentional error
```
Comment (zeevmoney) [inline sdk/typescript/test/e2e/dwallet-mpc.test.ts:96]:
```suggestion
```
??

**Review by zeevmoney — CHANGES_REQUESTED** [review]
see comments, also see the changes I made in the `move` and `ts` files.

**Review by zeevmoney — APPROVED** [review]
💯

**Pull request merged**